### PR TITLE
adding reduce capability to collections

### PR DESCRIPTION
**Pull request by rafaelmaza**
as in http://php.net/manual/en/function.array-reduce.php

### CODE REVIEW

File: lib/Doctrine/Common/Collections/Collection.php [hunk 1]
```diff
@@ -223,6 +223,16 @@ public function forAll(Closure $p);
...
+    
+    /**
+     * Applies the given function to each element in the collection and returns
+     * a single value.
+     *
+     * @param Closure $func
+     *
+     * @return Collection
+     */
+    public function reduce(Closure $func);
```
Comment (Ocramius) [inline lib/Doctrine/Common/Collections/Collection.php]: Interface additions are a BC break, sorry.

**Comment by stof** [comment]
@Ocramius should we add a warning about it in the Contributing file ? It may help reducing the number of PRs opened to add new methods

**Comment by mikeSimonson** [comment]
@rafaelmaza Thanks for your contribution. Sadly for technical reason this part of the ORM is really tightly coupled to our user and to the rest of the ORM. We really can't make any change that wouldn't be very Strickland backward compatible.

@stof We should definetly mention that in the contributing file.

**Comment by stevenwadejr** [comment]
Serious question: *why* and *how* is an **addition** of a method a breaking change and not backwards compatible? Changing or removing something, sure, but *adding* a method?

**Comment by stof** [comment]
@stevenwadejr this adds a method in an interface which can be (and *is*)  implemented in other packages than `doctrine/collections`. Adding a new method will mean that any other package now contain a broken class (because it misses the new method)

**Comment by stevenwadejr** [comment]
@stof Thanks. That makes perfect sense. 👍 I appreciate it.

**Comment by davidstanley01** [comment]
@stof Do those other packages only implement the `Collection` interface and not extend either the `ArrayCollection` or `AbstractLazyCollection` classes, too? If so, then I agree - this would be a breaking change. If they also extend those two classes, though, then this is _not_ a BC change as this PR (and #121) both implement the method.

**Comment by Ocramius** [comment]
@davidstanley01 we could make these changes if we had private interfaces, but we don't. As soon as an interface lands in a public package, BC concerns affect it.

**Comment by stof** [comment]
@stevenwadejr see https://symfony.com/bc The Symfony core team did a huge work about documenting exactly about what kind of changes are allowed or no regarding backward compatibility (many people forget about some.
Note that this is related to the BC guarantees we provide in the first section. If you want to turn some `No` into `Yes` there to be even stricter, it requires changing lots of `yes` into `No` about allowed changes.

@davidstanley01 they also implement the interface directly in some cases (which is perfectly valid usage of the package)

**Comment by davidstanley01** [comment]
Appreciate your time and the explanations. Thank you!

If you are open to enhancements for the next version, I can certainly help out. We use Doctrine heavily and adding these convenience methods would help out a bunch. 

Thanks, again.

**Comment by Ocramius** [comment]
@davidstanley01 I think that the collections API will be split out somehow: having the methods on the class itself is likely going to be a problem (this one, in particular) in the future too...

**Comment by davidstanley01** [comment]
@Ocramius Ok. At any rate, I'm happy to help when/if we get to that point. Cheers!

**Comment by stof** [comment]
@Ocramius especially for methods getting a callable as argument: there is no way to move this at DB-level for lazy collections, so have a utility class getting the collection as one of the argument of the method could make sense.

**Pull request closed without merging**